community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/SchemaRead.java

```diff
@@ -41,9 +41,19 @@
      *
      * @param label the index label
      * @param properties the index properties
-     * @return the IndexReference, or {@link CapableIndexReference#NO_INDEX} if such an index does not exist.
+     * @return the IndexReference, or {@link IndexReference#NO_INDEX} if such an index does not exist.
      */
-    CapableIndexReference index( int label, int... properties );
+    IndexReference index( int label, int... properties );
+
+    /**
+     * Acquire an index reference of the given {@code label} and {@code properties}. This method does not assert
+     * that the created reference points to a valid online index.
+     *
+     * @param label the index label
+     * @param properties the index properties
+     * @return a IndexReference for the given label and properties
+     */
+    IndexReference indexReferenceUnchecked( int label, int... properties );
```

Comment:
Do you mind sharing the thoughts behind this method? Is it "just" a factory method for `IndexReference`s so that we don't leak internal implementation details or is there something more intricate hiding here?

Reply:
It's just a factory method. It's needed because cypher looses the real index reference instance during planning by mapping to it's internal representation. When we come to execution we need again to provide an index reference to the kernel, but we don't want to pay the cost of validating the index existence again by various hash lookups.

We could get rid of most usages of this method also, but not all until `4.0`.